File: Algorithms/179.LargestNumber.java
public class Solution {
    public String largestNumber(int[] num) {
        List<Integer> new_num = new ArrayList<>();
        for (int i=0; i < num.length; i++) {
            new_num.add(num[i]);
        }
        Collections.sort(new_num, new cmp());
        if (new_num.get(0) == 0)
            return "0";
        String result = "";
        for (int i = 0; i < new_num.size(); i++) {
            result += new_num.get(i);
        }
        return result;
    }
    class cmp implements Comparator<Integer> {
        @Override
        public int compare(Integer x, Integer y) {
            long i = 10;
            while (i <= y) {
                i *= 10;
            }
            long xx = x*i + y;
            i = 10;
            while (i <= x) {
                i *= 10;
            }
            long yy = y * i + x;
            if (xx < yy)
                return 1;
            return -1;
        }
    }
}

File: Algorithms/cpp/179.LargestNumber.cpp
class Solution {
    public:
        string largestNumber(vector<int>& nums) {
            sort(nums.begin(), nums.end(), cmp);
            if (nums[0] == 0)
                return "0";
            string result = "";
            for (int i = 0; i < nums.size(); i++) {
                result += to_string(nums[i]);
            }
            return result;
        }
    private:
        bool static cmp(long x, long y) {
            long i = 10;
            while (i <= y) {
                i *= 10;
            }
            long xx = x*i + y;
            i = 10;
            while (i <= x) {
                i *= 10;
            }
            long yy = y * i + x;
            if (xx <= yy)
                return false;
            return true;
        }
};
